runtime/cudaq/qis/quantum_state.cpp

```diff
@@ -0,0 +1,113 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "quantum_state.h"
+#include "common/Logger.h"
+
+namespace cudaq {
+
+QuantumState::~QuantumState() {
+  if (!platformExecutionLog.empty()) {
+    // Flush any info log from the remote execution
+    printf("%s\n", platformExecutionLog.c_str());
+    platformExecutionLog.clear();
+  }
+
+  for (std::size_t counter = 0; auto &ptr : args)
+    deleters[counter++](ptr);
+
+  args.clear();
+  deleters.clear();
+}
+
+std::size_t QuantumState::getNumQubits() const {
+  throw std::runtime_error(
```

Comment:
Seems like all of these stubs are masking coding bugs until runtime and then just cause things to crash.

Can we just make `QuantumState` a pure interface and all of these methods pure virtual? That would force any derived classes to think it through and provide a proper implementation. An alternative would be to use the CRTP and the instantiation would require that the implementation class define all of the methods correctly.